Add HPA 'configurable tolerance' blog post (KEP-4951). (#50000)
kubernetes/website: content/en/blog/_posts/XXXX-XX-XX-hpa-configurable-tolerance.md

---

> layout: blog
> title: "Horizontal Autoscaling Configurable Tolerance"
> slug: hpa-configurable-tolerance
> date: XXXX-XX-XX
> author: "Jean-Marc François"

Reviewer: ```suggestion
layout: blog
title: "Kubernetes v1.33 and Horizontal Autoscaling Configurable Tolerance"
slug: hpa-configurable-tolerance
# after the v1.33 release, set a future publication date and remove the draft marker
# the release comms team can confirm which date has been assigned
#
# PRs to remove the draft marker should be opened BEFORE release day
draft: true 
math: true # for formulae
date: XXXX-XX-XX
author: "Jean-Marc François"
```

@jm-franc if you want to list a company affiliation, please do.
eg `author: Jean-Marc François (ACME Corp)`

---

> This post describes _Configurable Tolerance for Pod Horizontal Autoscaling_,

Reviewer: How about:
```suggestion
This post describes _configurable tolerance for horizontal Pod autoscaling_,
```

a new alpha feature first available in Kubernetes 1.33.

## What is it?

> Horizontal Pod Autoscaling (HPA) is a well-known Kubernetes feature that

Reviewer: ```suggestion
[Horizontal Pod autoscaling](/docs/tasks/run-application/horizontal-pod-autoscale/) (HPA) is a
well-known Kubernetes feature that
```
(aligning with our [style guide](https://kubernetes.io/docs/contribute/style/style-guide/))

allows your workload to automatically resize by adding or removing replicas
based on resource utilization.

To decide how many replicas a workload requires, users configure their HPA
with a metric (e.g. CPU utilization) and an expected value for this metric (e.g.
80%). The HPA updates the number of replica based on the ratio between the
current and desired metric value. (For example, if there are currently 100
replicas, the CPU utilization is 88%, and the desired utilization is 80%, the

> HPA will ask for `100 * (88/80)` replicas).

Reviewer: We prefer rendered formulae; see https://kubernetes.io/docs/contribute/style/style-guide/#formulae-and-equations

You can also use `×` rather than `*` here, but a formula is what we recommend.

In order to avoid replicas being created or deleted whenever a small metric

> fluctuation occurs, Kubernetes require that the current and desired metrics

Reviewer: ```suggestion
fluctuation occurs, Kubernetes applies a form of hysteresis: it only changes the number of replicas
when the the current and desired metric values
```

differ by more than 10%.

> This tolerance of 10% is cluster-wide and typically cannot be fine-tuned. It's

Reviewer: ```suggestion
This default tolerance of 10% is cluster-wide; in older Kubernetes releases, it could not be fine-tuned. It's
```

a suitable value for most usage, but too coarse for large deployments, where a
10% tolerance represents tens of pods. As a result,

> [users have long asked](https://github.com/kubernetes/kubernetes/issues/116984)

Reviewer: (nit)
``suggestion
users have long [asked](https://github.com/kubernetes/kubernetes/issues/116984)
```

to be able to tune this value.

> Thanks to the _Configurable Tolerance for Pod Horizontal Autoscaling_ feature,

Reviewer: ```suggestion
In Kubernetes v1.33,
``

this is now possible.

## How do I use it?

Just add the tolerance you want an HPA to use to your `HorizontalPodAutoscaler`

> resource.

Reviewer: This sounds wrong. Do you not also need to enable a feature gate?

Also, per the style guide:
- write "object" not "resource"
- don't use backticks here

Author: Done (I've moved the paragraph mentioning which feature flag gates this).

Tolerances appear under the `spec.behavior.scaleDown` and
`spec.behavior.scaleUp` fields and can thus be different for scale up and scale
down. A typical usage would be to specify a small tolerance on scale up (to
react quickly to spikes), but lower on scale down (to avoid adding and removing
replicas too quickly in response to small metric fluctuations).

For example, an HPA with a tolerance of 5% on scale-down, and no
tolerance on scale-up, would look like the following:

```yaml

> apiVersion: autoscaling/v2beta2

Reviewer: :warning: **Don't recommend using deprecated APIs (nor imply a recommendation)**

```suggestion
apiVersion: autoscaling/v2
```

Reviewer: Easy slip to make, but in our official articles, we need to be careful about what we imply.

kind: HorizontalPodAutoscaler
metadata:
  name: my-app

> spec:

Reviewer: Make it obvious that this is not a complete HorizontalPodAutoscaler manifest; for example:
```suggestion
spec:
   …
```

  behavior:
    scaleDown:
      tolerance: 0.05
    scaleUp:
      tolerance: 0
```

Note: This feature is in alpha in Kubernetes 1.33, gated by the

> `HPAConfigurableTolerance` flag.

Reviewer: Mention this earlier!

---
layout: blog
title: "Kubernetes v1.33 and Horizontal Autoscaling Configurable Tolerance"
slug: hpa-configurable-tolerance
# after the v1.33 release, set a future publication date and remove the draft marker
# the release comms team can confirm which date has been assigned
#
# PRs to remove the draft marker should be opened BEFORE release day
draft: true
math: true # for formulae
date: XXXX-XX-XX
author: "Jean-Marc François (Google)"
---

This post describes _configurable tolerance for horizontal Pod autoscaling_,
a new alpha feature first available in Kubernetes 1.33.

## What is it?

> [Horizontal Pod autoscaling](/docs/tasks/run-application/horizontal-pod-autoscale/)

Reviewer 2: I haven't seen this lower case "autoscaling" when other "Horizontal" and "Pod" are capitalised. Given this blog is going to talk about HPA more extensively, it may be more fitting to just use "Horizontal Pod Autoscaling"?

automatically resize by adding or removing replicas based on resource
utilization.

To decide how many replicas a workload requires, users configure their HPA
with a metric (e.g. CPU utilization) and an expected value for this metric (e.g.
80%). The HPA updates the number of replica based on the ratio between the
current and desired metric value. (For example, if there are currently 100
replicas, the CPU utilization is 88%, and the desired utilization is 80%, the
HPA will ask for \\(100 \times (88/80)\)) replicas).

In order to avoid replicas being created or deleted whenever a small metric
fluctuation occurs, Kubernetes applies a form of hysteresis: it only changes the
number of replicas when the the current and desired metric values differ by more
than 10%.

This default tolerance of 10% is cluster-wide; in older Kubernetes releases, it
could not be fine-tuned. It's a suitable value for most usage, but too coarse
for large deployments, where a 10% tolerance represents tens of pods. As a
result, users have long
[asked](https://github.com/kubernetes/kubernetes/issues/116984) to be able to
tune this value.

In Kubernetes v1.33, this is now possible.

## How do I use it?

> Enable the `HPAConfigurableTolerance` feature flag in your Kubernetes 1.33

Reviewer: Kubernetes uses the term "feature gate" to describe feature flags.

```suggestion
Enable the `HPAConfigurableTolerance` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/) in your Kubernetes 1.33
```

Author: Done, thank you!

> Enable the `HPAConfigurableTolerance` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/) in your Kubernetes 1.33
> cluster, then add the tolerance you want an HPA to use to your
> HorizontalPodAutoscaler object.

Reviewer 2: Just minor wording change suggestion!

```suggestion
After enabling the `HPAConfigurableTolerance` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/) in your Kubernetes v1.33
cluster, you can add your desired tolerance for your
HorizontalPodAutoscaler object.
```

Author: Done, thank you.

> Tolerances appear under the `spec.behavior.scaleDown` and
> `spec.behavior.scaleUp` fields and can thus be different for scale up and scale
> down. A typical usage would be to specify a small tolerance on scale up (to
> react quickly to spikes), but lower on scale down (to avoid adding and removing
> replicas too quickly in response to small metric fluctuations).

Reviewer 2: If I'm understanding this correctly, I think you meant to say "higher" on scale down? The tolerance being high would result in replicas to get removed only slowly?

```suggestion
Tolerances appear under the `spec.behavior.scaleDown` and
`spec.behavior.scaleUp` fields and can thus be different for scale up and scale
down. A typical usage would be to specify a small tolerance on scale up (to
react quickly to spikes), but higher one on scale down (to avoid adding and removing
replicas too quickly in response to small metric fluctuations).
```

Author: Yes, thank you! Done.

For example, an HPA with a tolerance of 5% on scale-down, and no tolerance on
scale-up, would look like the following:

```yaml
apiVersion: autoscaling/v2
kind: HorizontalPodAutoscaler
metadata:
  name: my-app
spec:
  ...
  behavior:

>     scaleDown:

Reviewer 3: I hope that my comment is clear and that I didn't misunderstood the implementation.

>     scaleDown:

Author: I've added this example (slightly adapted from a comment above): https://github.com/kubernetes/website/pull/50000/commits/30fc1259ee4488f2cf8f4ee625ccce32d0196bc0  I hope that's ok but let me know if you were thinking of something different.
